### PR TITLE
Add additional source suppoprt to app.mk

### LINKED ISSUE #4955: Extend app.mk to allow additional object files
For including external application codes it would be convenient to have an ```ADDITIONAL_APP_OBJECTS``` and an ```ADDITIONAL_APP_DEPS``` variable that is used by ```app.mk``` to construct teh app dependencies and object files to be built.

### PR DESCRIPTION
**Pull request by dschwen**
Add support for additional source files to `app.mk`. This is used by the upcoming Magpie app to compile in third-party apps.

Closes #4955.

### CODE REVIEW

**Comment by permcody** [comment]
Is this ready to go or do you think you have more work to get this all working with apps?

**Comment by dschwen** [comment]
This is ready to go and confirmed to work for building magpie/spparks

File: framework/build.mk [hunk 1]
```diff
@@ -29,7 +29,7 @@ libmesh_CC       ?= $(shell METHOD=$(METHOD) $(libmesh_config) --cc)
 libmesh_F77      ?= $(shell METHOD=$(METHOD) $(libmesh_config) --fc)
 libmesh_F90      ?= $(shell METHOD=$(METHOD) $(libmesh_config) --fc)
 libmesh_INCLUDE  := $(shell METHOD=$(METHOD) $(libmesh_config) --include)
-libmesh_CPPFLAGS := $(shell METHOD=$(METHOD) $(libmesh_config) --cppflags)
+libmesh_CPPFLAGS := $(shell METHOD=$(METHOD) $(libmesh_config) --cppflags) $(ADDITIONAL_CPPFLAGS)
```
Comment (permcody) [inline framework/build.mk]: I think I'd rather keep the libMesh flags pristine. If you need a new variable add it to the rule itself.

Comment (dschwen) [inline framework/build.mk]: That's what I did before, but it appears in 6 or seven spots :-/. But I can change it.

Comment (permcody) [inline framework/build.mk]: Yeah - I realize that's not ideal either... We probably should rename all of those variables to moose_FOO and do what we want with them. That'll be a larger cleanup task in the future.

**Pull request merged**